### PR TITLE
Issue 50138: Switch assay XLSX parsing to async streaming

### LINKED ISSUE #762: Importing giant assay XSLX file
[Issue](https://www.labkey.org/home/Developer/issues/Secure/issues-details.view?issueId=50138)

### Tasks
- [x] Multi-threaded parsing for XLSX - 3
- [x] Stream rows from ExcelLoader - 2
- [x] Stream rows through assay import - 3
- [x] Stream rows through transform scripts - 3
- [x] Stream rows for post-import - 3
- [x] Test fixes  
- [x] manual test plan - 1
- [x] manual testing - 2 @labkey-matthewb  
- [x] manual testing - 2 @labkey-chrisj  
- [x] Automated test coverage for .xlsx @labkey-chrisj 
  - Update existing tests to use .xlsx files and ensure both .xls and .xlsx variants are covered.
  - [x] New import/export coverage for 'worst-case' xlsx data
- [x] code review - 1 @labkey-matthewb 
- [x] Fix Luminex tests @labkey-jeckels https://teamcity.labkey.org/buildConfiguration/LabkeyTrunk_ModuleSuites_LuminexPostgres/3073010?hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildDeploymentsSection=false&expandBuildChangesSection=true&expandBuildTestsSection=true
- [x] Fix other tests https://teamcity.labkey.org/buildConfiguration/LabkeyTrunk_DailyPostgres_2/3078043?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildTestsSection=true @labkey-jeckels 
- [x] teamcity review and merge - 1

### Risk Assessment
_Identify potential risks and propose mitigation strategies. Each team member should contribute by highlighting any risks they foresee, considering business, user, project management, technical, and quality assurance aspects._

- Parsing changes: deadlocks, unclosed resources (especially in error conditions)
  - Impact: High
  - Likelihood: Medium/Low
  - Mitigation: Unit tests
- Risk: Mangled handling of edge cases (zero-row imports, required field checks, import aliases, etc). Possibility of data loss if columns aren't mapped correctly.
  - Impact: Medium
  - Likelihood: Medium
  - Mitigation: Manual test, review automated coverage and consider bolstering
- Risk: Mangled handling of post-import processing (provenance, plate metadata)
  - Impact: Medium
  - Likelihood: Medium
  - Mitigation: Manual test, review automated coverage and consider bolstering
- Risk: Mangled handling in other assay types
  - Impact: High
  - Likelihood: Low
  - Mitigation: Manual test, review automated coverage and consider bolstering

### Manual Test Plan
- Generally, be on the lookout for lingering threads with a name starting with `Async XLSX parser:`. These shouldn't stick around longer than the import is actively running. You can check via Running Threads in the Admin Console or the debugger.
- Standard assay
  - Import small/large XLSX files with expected structure (happy path)
    - Resolve fields via name, labels, import aliases, and property URIs
  - Verify zero row files give helpful error (consistently with 24.3)
  - Verify required fields get enforced with helpful error
  - Verify provenance is connected correctly (has automated coverage that I broke and then fixed)
  - Verify plate metadata works as before (note: Josh is unfamiliar with this area)
  - Verify transform scripts can transform data (note: currently loads all transformed rows into memory at once)
  - Verify basic import works for TSV/CSV as well (don't need to duplicate full XLSX testing)
  - Verify SaveBatch API works
- Luminex assay, basic use (has automated coverage)
- NAb assay, basic use (has automated coverage)
- Viability assay, basic use (has automated coverage)

### PR DESCRIPTION
**Pull request by labkey-jeckels**
#### Rationale
Excel files can be very large. It's better to stream their data during import instead of holding it all in memory at once.

#### Changes
- Use a multi-threaded parsing approach, letting the SAX parser fill a blocking queue and having the main thread consume data rows
- Process data as a stream of `Map<String, Object>` instead of `List<Map<String, Object>>`

### CODE REVIEW

File: api/src/org/labkey/api/iterator/ValidatingDataRowIterator.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+public interface ValidatingDataRowIterator extends Closeable
```
Comment (labkey-matthewb) [inline api/src/org/labkey/api/iterator/ValidatingDataRowIterator.java:21]: Just curious: why not implement Iterator<Map<String,Object>>?

Comment (labkey-jeckels) [reply api/src/org/labkey/api/iterator/ValidatingDataRowIterator.java:21]: That's where I started but eventually needed to throw `ValidationException` from `next()`. I will add a comment explaining.

Comment (labkey-matthewb) [reply api/src/org/labkey/api/iterator/ValidatingDataRowIterator.java:21]: Got it.

File: assay/src/org/labkey/assay/plate/AssayPlateMetadataServiceImpl.java [hunk 2]
```diff
@@ -144,36 +144,44 @@ public void addAssayPlateMetadata(
...
-                if (tableInfo != null)
+                if (!jsonData.isEmpty())
                 {
-                    QueryUpdateService qus = tableInfo.getUpdateService();
-                    BatchValidationException errors = new BatchValidationException();
-
-                    qus.insertRows(user, container, jsonData, errors, null, null);
-                    if (errors.hasErrors())
+                    AssayProtocolSchema schema = provider.createProtocolSchema(user, container, protocol, null);
+                    TableInfo tableInfo = schema.createTable(TSVProtocolSchema.PLATE_DATA_TABLE, null);
```
Comment (labkey-matthewb) [inline assay/src/org/labkey/assay/plate/AssayPlateMetadataServiceImpl.java:162]: existing code, but I wonder why createTable() instead of the usual getTable()

File: api/src/org/labkey/api/iterator/WrappingCloseableIterator.java [hunk 3]
```diff
@@ -51,7 +54,11 @@ public void remove()
     }
 
     @Override
-    public void close()
+    public void close() throws IOException
```
Comment (labkey-matthewb) [inline api/src/org/labkey/api/iterator/WrappingCloseableIterator.java:57]: The pattern I've used in DataIterator land is to do "static CloseableIterator wrap(Iterator)"  So the wrapping can be skipped if the passed in parameter implements CloseableIterator.

File: study/test/src/org/labkey/test/tests/study/AutoLinkToStudyTest.java [hunk 4]
```diff
@@ -129,7 +129,7 @@ public void testLinkedColumnNotDisplayedCase()
                 .setAutoLinkTarget("(Data import folder)")
                 .clickSave();
 
-        File runFile = new File(TestFileUtils.getSampleData("AssayImportExport"), "GenericAssay_Run2.xls");
+        File runFile = TestFileUtils.getSampleData("AssayImportExport/GenericAssay_Run2.xlsx");
```
Comment (labkey-matthewb) [inline study/test/src/org/labkey/test/tests/study/AutoLinkToStudyTest.java:132]: 8\)

**Review by labkey-matthewb — APPROVED** [review]
DataIteratorBuilder > Supplier<ValidatingDataRowIterator>
